patches/libtock-rs/0001-alloc-init-feature.patch

```diff
@@ -0,0 +1,121 @@
+@@ -10,6 +10,8 @@ core::arch::global_asm!(include_str!("asm_arm.s"));
+ #[cfg(target_arch = "riscv32")]
+ core::arch::global_asm!(include_str!("asm_riscv32.s"));
+ 
++static APP_HEAP_SIZE: Option<&'static str> = option_env!("APP_HEAP_SIZE");
++
+ /// `set_main!` is used to tell `libtock_runtime` where the process binary's
+ /// `main` function is. The process binary's `main` function must have the
+ /// signature `FnOnce() -> T`, where T is some concrete type that implements
+@@ -83,6 +85,8 @@ extern "C" fn rust_start() -> ! {
+     extern "Rust" {
+         fn libtock_unsafe_main() -> !;
+         static rt_header: RtHeader;
++        #[cfg(feature = "alloc_init")]
++        fn libtock_alloc_init(heap_bottom: *mut u8, heap_size: usize);
+     }
+ 
+     // TODO: Implement a safe memop API in libtock_platform and migrate these
+@@ -92,16 +96,46 @@ extern "C" fn rust_start() -> ! {
+     // impact the execution of this process.
+     #[cfg(not(feature = "no_debug_memop"))]
+     unsafe {
++        // specify the top of the application stack which grows downwards
+         TockSyscalls::syscall2::<{ syscall_class::MEMOP }>([
+             10u32.into(),
+             rt_header.stack_top.into(),
+         ]);
++
++        // specify the start of the application heap which grows upwards
+         TockSyscalls::syscall2::<{ syscall_class::MEMOP }>([
+             11u32.into(),
+             rt_header.initial_break.into(),
+         ]);
+     }
+ 
++    #[cfg(feature = "alloc_init")]
++    {
++        let app_heap_size: usize = match APP_HEAP_SIZE {
++            Some(var) => var
++                .parse()
++                .ok()
++                .expect("could not parse APP_HEAP_SIZE as usize!"),
++            None => 9000,
++        };
++
++        // the heap starts after the `bss` section
++        let app_heap_bottom = unsafe { rt_header.bss_start.add(rt_header.bss_size) };
++
++        assert_eq!(app_heap_bottom.addr(), unsafe {
++            rt_header.bss_start.addr() + rt_header.bss_size
++        });
++
++        let app_heap_end = unsafe { app_heap_bottom.add(app_heap_size) };
++
++        unsafe {
++            // tell the kernel the new app heap break (which is the upper address bound of the process)
++            TockSyscalls::syscall2::<{ syscall_class::MEMOP }>([0u32.into(), app_heap_end.into()]);
++
++            libtock_alloc_init(app_heap_bottom, app_heap_size);
++        }
++    }
++
+     // Safety: libtock_unsafe_main is defined by the set_main! macro, and its
+     // signature matches the signature in the `extern` block in this function.
+     unsafe {
+diff --git a/rust-toolchain b/rust-toolchain
+index e911a96..d00feab 100644
+--- a/rust-toolchain
++++ b/rust-toolchain
+@@ -1,7 +1,7 @@
+ [toolchain]
+ # See https://rust-lang.github.io/rustup-components-history/ for a list of
+ # recently nightlies and what components are available for them.
+-channel = "nightly-2022-06-10"
++channel = "nightly-2022-10-22"
```

Comment:
Can we go newer than this?

Reply:
Probably. This specific pinned version is just because I started around this time. How far ahead do you want to go with the nightlies? Is there some specific feature you want to include?

Reply:
Just wanted to understand if there is any blocker, and how much new freedom we have.